### ISSUE TITLE
Impact of input image size on output quality

### ISSUE THREAD

**Issue by sohrab-m** [piddnad/DDColor#54]
Hi, 

We have been playing around with the inference of the HF Model, using a simple inference script: 
```python
from inference.colorization_pipeline import ImageColorizationPipeline
import cv2
import os
from tqdm import tqdm

input_size = 256 
test_path = os.path.join("assets", "test_images")
results_path = os.path.join(f"results-{input_size:d}")    

os.makedirs(results_path, exist_ok=True)
img_list = os.listdir(test_path)

colorizer = ImageColorizationPipeline(model_path=os.path.join("pretrain", "ddcolor_paper.pt"), input_size=input_size, model_size='large')

for name in tqdm(img_list):
        img = cv2.imread(os.path.join(test_path, name))
        image_out = colorizer.process(img)
        cv2.imwrite(os.path.join(results_path, name), image_out)
``` 
Two things; 
1. By changing the `input_size` to `128, 256, 320, 480, 512, 1024`, the model outputs significantly different colorizations. For instance, an image  resized to 256 might remain mostly black and white, while the same image resized to 512 might get nicely colorized comparable to those in the paper. You can see a qualitative example of this in the three examples below.
* There does not seem to be a clear trend (e.g., larger input size better or worse, or aspect ratio affecting this phenomenon) with the limited testing we have done so far.

2. On many occasions, there will be small artifacts of color patches added to the colorized output, usually a red/blue or green/blue combinations. You can see an example of this on the second image below, with `input_size = 320`. Again, we have not found a clear trend behind these patches.

<p align="center">
  <div style="display: inline-block; text-align: center; margin: 30px;">
    <img src="https://github.com/user-attachments/assets/873cd808-ed4f-45cb-ad32-cea52c8f4599" alt="Image 1" width="50%" />
    <div>Input size = 256</div>
  </div>
  <div style="display: inline-block; text-align: center; margin: 30px;">
    <img src="https://github.com/user-attachments/assets/b40032dc-611e-4983-a56f-541fa8a4d730" alt="Image 2" width="50%" />
    <div>Input size = 320</div>
  </div>
  <div style="display: inline-block; text-align: center; margin: 30px;">
    <img src="https://github.com/user-attachments/assets/7ef37bef-483a-480a-8b29-cc6e47ed220f" alt="Image 3" width="50%" />
    <div>Input size = 480</div>
  </div>
</p>

Have you encountered these before, and if so, do you know how we can address them?

**Comment by piddnad** [comment]
Hi @sohrab-m , thank you for sharing your observations. While I don’t have access to the original experimental records, I can offer some insights based on memory.

1. Input resolution: The ImageNet training dataset includes some black-and-white historical photographs. As a result, the model may have learned to identify certain inputs as “possibly old photographs” at specific resolutions (i.e. 256px for ddcolor_paper model), which could lead to a tendency to preserve the grayscale output.

2. Color patch artifacts: These might result from the colorfulness loss used in training. While it generally produces more vibrant results, it can occasionally introduce artifacts in specific cases.

Have you tried using other models, such as ddcolor_artistic.pth (without colorfulness loss) or ddcolor_modelscope.pth (see https://github.com/piddnad/DDColor/blob/master/MODEL_ZOO.md)? These might give you different results and address some of the issues you’ve observed.

**Comment by sohrab-m** [comment]
Hi @piddnad, thank you for the response. We have tried the modelscope one with similar results -- but that model also uses colorfulness loss. We will try models without the colorfulness loss to see what happens.

**Comment by alexaex** [comment]
@sohrab-m @piddnad 

I hope this can help you.

**According to my experiments, I am for sure the cause of these color artifacts. The issue stems from chrominance value overflow in the color space.**

the chrominance:
![Image](https://github.com/user-attachments/assets/8a4d2f01-952f-42e3-b2d8-f8e282e7b289)
the output of DDColor:
![Image](https://github.com/user-attachments/assets/86a574c5-9dba-41a3-9f6a-72a34be10bbf)
the a-b values in the lower right corner approach extreme negative values between -170 and -180. 
![Image](https://github.com/user-attachments/assets/087ad160-16c9-48b3-8110-7399de12fd08)

I have also been troubled by this problem for a long time. Initially, I suspected there might be issues with my implementation of DDColor. However, when I switched to using the official code, the same problem persisted. After reviewing OpenCV's documentation on the CIE-LAB color space range, I confirmed my suspicions by examining the chrominance values. 

**According to OpenCV documentation, the valid range for chrominance values in the CIE-LAB color space is limited to -128 to 127.**

you can use the code below to verify it.
  @torch.no_grad()
    def process(self, img):
        height, width = img.shape[:2]

        img = (img / 255.0).astype(np.float32)
        orig_l = cv2.cvtColor(img, cv2.COLOR_BGR2Lab)[:, :, :1]  # (h, w, 1)

        # Resize and convert image to grayscale
        img_resized = cv2.resize(img, (self.input_size, self.input_size))
        img_l = cv2.cvtColor(img_resized, cv2.COLOR_BGR2Lab)[:, :, :1]
        img_gray_lab = np.concatenate((img_l, np.zeros_like(img_l), np.zeros_like(img_l)), axis=-1)
        img_gray_rgb = cv2.cvtColor(img_gray_lab, cv2.COLOR_LAB2RGB)

        tensor_gray_rgb = torch.from_numpy(img_gray_rgb.transpose((2, 0, 1))).float().unsqueeze(0).to(self.device)
        output_ab = self.model(tensor_gray_rgb).cpu()  # (1, 2, self.input_size, self.input_size)

        # Resize output and concatenate with original L channel
        output_ab_resized = F.interpolate(output_ab, size=(height, width))[0].float().numpy().transpose(1, 2, 0)
        import sys
        import numpy
        numpy.set_printoptions(threshold=sys.maxsize)
        print(output_ab_resized[..., 0])
        print(output_ab_resized[..., 1])
        chrominance = np.concatenate([np.zeros_like(output_ab_resized[..., 0:1]), output_ab_resized], axis=-1)
        chrominance = cv2.cvtColor(chrominance, cv2.COLOR_Lab2RGB)
        import matplotlib.pyplot as plt
        plt.imshow(chrominance)
        plt.show()
        output_lab = np.concatenate((orig_l, output_ab_resized), axis=-1)
        output_bgr = cv2.cvtColor(output_lab, cv2.COLOR_LAB2BGR)

        output_img = (output_bgr * 255.0).round().astype(np.uint8)
        return output_img
